### PR TITLE
Foursquare provider

### PR DESCRIPTION
**Pull request by rjperes**

### CODE REVIEW

**Comment by rjperes** [comment]
Hi, @jerriep and @ByteBlast! This is my first pull request here... hope it gets accepted!
BTW, congrats on your project! ;-)

**Comment by AlexArchive** [comment]
Hi @rjperes. Many thanks for your contribution. I am checking this out now.

**Comment by rjperes** [comment]
@ByteBlast! Thanks! BTW, where can I ask for help/directions on troubleshooting the Yammer provider? Shall I use "Issues"?

**Comment by AlexArchive** [comment]
Yep - please create an issue.

File: Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+				var oauth2Token = JObject.Parse(oauthTokenResponse);
+				var accessToken = oauth2Token["access_token"].Value<String>();
+
+				if (String.IsNullOrWhiteSpace(accessToken) == true)
+				{
+					this._logger.WriteWarning("Access token was not found");
+					return new AuthenticationTicket(null, properties);
+				}
+
+				var graphResponse = await this._httpClient.GetAsync(GraphApiEndpoint + "?oauth_token=" + Uri.EscapeDataString(accessToken), Request.CallCancelled);
```
Comment (AlexArchive) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs]: Hm. On my computer running on `localhost` this line results in a 410 "Gone". This causes the subsequent statement (`EnsureSuccessStatusCode`) to throw an exception. I assume this does not happen on your machine...?

Comment (rjperes) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs]: No, it doesn't, but I never ran it in localhost... I use 127.0.0.1.xip.io and configure IIS Express accordingly. Can you try it with a non-local address, please?

**Comment by AlexArchive** [comment]
Unfortunately this provider does not work on my `localhost`. I commented on the line I think is causing problems. Can you please investigate?

There are some small formatting issues, and some convention issues (we use `string` not `String`) but I can fix those for you.

**Comment by rjperes** [comment]
OK, but _should_ it run on localhost? Some other providers - Microsoft, etc - require a fully qualified domain name, I suppose Foursquare could too?

**Comment by rjperes** [comment]
It does work with localhost, just checked! But you have to add it to the list of redirect urls in developer.foursquare.com. In my application, I have "http://127.0.0.1.xip.io:32851/, http://localhost:32851".

**Comment by AlexArchive** [comment]
~~Well. I know the Microsoft provider flat out prevents you from inputting a loop back address such as `localaddress` - Foursquare does not.~~

~~The provider _does_ attain an access token  - It is the `user` api that is resulting in an error.~~

Hold on.

**Comment by AlexArchive** [comment]
Sorry. But where **specifically** should I register the redirect url?

I already registered an application with the correct redirect url: 



Must I register the redirect url separately?

Understand that the redirect url does get invoked by Foursquare. It is when I access `user/self` that the 410 is returned.

**Comment by rjperes** [comment]
BTW, I replaced "String" for "string", "Object" for "object", etc, but then I noticed you have these all over the place! :-D

**Comment by AlexArchive** [comment]
Hehe. You are right - my bad. I will be changing them all soon. See #67.

**Comment by rjperes** [comment]
For the redirect URL, just use http://localhost:1729, works here, except it's a different port. Use the same application, of course, otherwise you have to change id and secret.

**Comment by AlexArchive** [comment]
I really do not think there is anything wrong with the redirect url. Refer to this [screenshot](http://i.imgur.com/Sk1aT82.png). The return path _is_ invoked.

**Comment by AlexArchive** [comment]
Next refer to this [screenshot](http://i.imgur.com/jYCpGKF.png). We have an access token and we use it to communicate with the `user/self` end-point but that request returns a 410.

**Comment by AlexArchive** [comment]
I think you need to add a version paramter like: `&v=20150318`. 

Here is the response body:

``` json
{  
   "meta":{  
      "code":410,
      "errorType":"param_error",
      "errorDetail":"The Foursquare API no longer supports requests that do not pass in a version parameter. For more details see https://developer.foursquare.com/overview/versioning"
   },
   "response":{  

   }
}
```

**Comment by rjperes** [comment]
Well, I get a 200, and it works perfectly, using localhost!
What else could be different?

**Comment by rjperes** [comment]
Works with 20130910 and also with 20150318.

**Comment by AlexArchive** [comment]
You did not include the version number in this pull request. 

You access

```
https://api.foursquare.com/v2/users/self?oauth_token=ABC
```

When you should access

```
https://api.foursquare.com/v2/users/self?oauth_token=ABC&v=20150318
```

See:

``` C#
var graphResponse = await this._httpClient.GetAsync(GraphApiEndpoint + "?oauth_token=" + Uri.EscapeDataString(accessToken), Request.CallCancelled);
```

I do not know why this works on your machine unless this pull request is missing local changes.

**Comment by rjperes** [comment]
Arrrggghhh... sorry... probably last minute changes/cleanup... will update it in a minute, ok? Sorry for this...

**Comment by AlexArchive** [comment]
No problem bud. 

While you're at it, could you please:
1. Add your provider to the demo project (comment out the line and leave the using directive)
2. Run `Control + K, D` (Visual Studio shortcut) on every file. Whilst the formatting looks fine in Visual Studio, it does not on GitHub. [See](https://github.com/owin-middleware/OwinOAuthProviders/pull/69/files).

Brb. Dinner.

**Comment by rjperes** [comment]
New version waiting for you for dessert!

File: Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+				context.Identity = new ClaimsIdentity(
+					new[]
+					{
+						new Claim(ClaimTypes.NameIdentifier, context.Id, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim(ClaimTypes.Name, context.Name, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:id", context.Id, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:name", context.Name, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:email", context.Email, XmlSchemaString, this.Options.AuthenticationType),
+						new Claim("urn:foursquare:twitter", context.Twitter, XmlSchemaString, this.Options.AuthenticationType)
```
Comment (AlexArchive) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs]: It is possible that the authenticated user does not have a Twitter account or public email, and so on. You need to surround **every** claim in an `if` statement like [this](https://github.com/owin-middleware/OwinOAuthProviders/blob/master/Owin.Security.Providers/GitHub/GitHubAuthenticationHandler.cs#L98). If you do not, an exception will be thrown for users who are "missing" values.

**Comment by AlexArchive** [comment]
Hello. I am back.

You need to surround each claim in an `if` statement before I can pull this in. See my comment for more information.

I do not think you corrected the formatting issue either :cry: - the formatting is still a bit messed up here. [See](https://github.com/owin-middleware/OwinOAuthProviders/pull/69/files).

File: Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs (hunk 2 above)
Comment (rjperes) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs]: Right

**Comment by rjperes** [comment]
Ok... Give-me a moment...

**Comment by AlexArchive** [comment]
Yep. No problem. Many thanks for working with me here @rjperes :smile:.

**Comment by rjperes** [comment]
:-)
You know, I did this because I plan to use your library and I need Foursquare auth... maybe I will submit pull requests for others as well, you will judge if they are relevant or not.
BTW, the OAuth for ASP.NET site could use some more info... but I guess it's @jerriep's, not yours, right?

**Comment by AlexArchive** [comment]
Woohoo! I can finally log on - nice work.

Yeah. @jerriep is the sole owner of http://www.oauthforaspnet.com/. He tells me he is working on some updates to the site - I am sure he will be interested to hear your feedback :smile:. You can ping him on Twitter [@jerriepelser](https://twitter.com/jerriepelser).

**Comment by rjperes** [comment]
Thanks!

File: Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+				if (string.IsNullOrEmpty(extra.RedirectUri) == true)
+				{
+					extra.RedirectUri = currentUri;
+				}
+
+				// OAuth2 10.12 CSRF
+				this.GenerateCorrelationId(extra);
+
+				// OAuth2 3.3 space separated
+				var scope = string.Join(" ", this.Options.Scope);
```
Comment (AlexArchive) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs:178]: This variable is not assigned. If it is not needed can you please remove it and the `Scope` property from the options class. If it is needed, please update the query string. Thank you.

File: Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+				var state = this.Options.StateDataFormat.Protect(extra);
+
+				var authorizationEndpoint = AuthorizationEndpoint +
+						"?client_id=" + Uri.EscapeDataString(this.Options.ClientId) +
+						"&response_type=code" +
+						"&redirect_uri=" + Uri.EscapeDataString(redirectUri) +
+						"&state=" + Uri.EscapeDataString(state);
+
+				this.Response.StatusCode = (int)HttpStatusCode.Moved;
+				this.Response.Headers.Set("Location", authorizationEndpoint);
```
Comment (AlexArchive) [inline Owin.Security.Providers/Foursquare/FoursquareAuthenticationHandler.cs]: I think you can replace lines 188 and 189 with 

``` C#
Response.Redirect(authorizationEndpoint);
```

(No need to worry about this one if you do not want to.)

**Comment by AlexArchive** [comment]
We are almost done. I made a couple more comments.

Have you noticed how when you are redirected the URL has a fragment identifier in it? 

```
http://localhost:1729/#_=_
```

No need to worry about this today but I am just wondering, do you know why? It would be nice to remove it.

**Comment by rjperes** [comment]
Here you go. I kept the scope for coherence with the other providers, even if Foursquare doesn't use them.

**Comment by AlexArchive** [comment]
I will bring this in now as a work in in progress. It is good work but before I bump the version number I want to address some or all of these problems:
- Url fragment shouldn't be there
- The context class is not modeled very well in my opinion
- I think some comments are missing
- We can probably remove scope. 

Thank you for all your work today @rjperes. 

Sorry. I forgot to respond to:

> You know, I did this because I plan to use your library and I need Foursquare auth... maybe I will submit pull requests for others as well, you will judge if they are relevant or not.

I recommend that you create an issue telling us what provider you want to implement before doing any work - we might not want it.

**Comment by rjperes** [comment]
OK, what can I do to improve it? Comments are easy, same for scope, but what about the Url fragment and the context class? You know, I did copy most of it from other providers of yours.

**Comment by AlexArchive** [comment]
I re-read my message and it sounded a bit and mean. Sorry. I really appreciate your contribution and I owe it to you to write a more elaborate description of the issues - I will create a descriptive issue tomorrow.

**Comment by rjperes** [comment]
Not at all! :-D
I'll do my best to get this right and available in NuGet! ;-)

**Comment by jerriep** [comment]
@rjperes We are looking for help moving to ASP.NET 5. Please read http://www.jerriepelser.com/blog/calling-contributors-for-oauth-providers

**Pull request merged**